### PR TITLE
Fix knitr and testing toolchain

### LINKED ISSUE #32: Markdown integration somewhat broken
I'm experiencing problems with the Markdown integration. For some reason the Rust return value is not captured anymore, and neither is any output generated by Rust. 

For example, in the vignette I get this output (`NULL` instead of 35):
Screen Shot 2021-02-03 at 1 14 25 PM
or this output (text printed by Rust is not output):
Screen Shot 2021-02-03 at 1 15 00 PM

It looks like this on the website currently, so this must have been broken for a while:
https://extendr.github.io/rextendr/articles/rmarkdown.html

It worked in the past. Maybe something has changed in knitr package. As far as I can tell, it's not related to the `rust_eval()` function that evaluates the code. That function works fine.

### PR DESCRIPTION
**Pull request by clauswilke**
Fix knitr integration. Closes #32.

Also adds knitr unit tests and fixes setting the test toolchain.

### CODE REVIEW

**Comment by clauswilke** [comment]
@Ilia-Kosenkov Can you check whether setting the toolchain via env var still works for you in this branch? It should, but the knitr test I added may fail. Not sure.

File: R/eval.R [hunk 1]
```diff
@@ -32,6 +32,9 @@ rust_eval <- function(code, env = parent.frame(), ...) {
   # make sure code is given as a single character string
   code <- glue::glue_collapse(code, sep = "\n")
 
+  # define to make R code check happy; is not used
+  rextendr_rust_eval_fun <- function() "decoy function; should never be called."
```
Comment (Ilia-Kosenkov) [inline R/eval.R:36]: Perhaps this message should be put into `stop()`?
If it gets called (somehow), the output may be swallowed, but an explicit error will indicate something is not right.

**Comment by Ilia-Kosenkov** [comment]
> 
> 
> @Ilia-Kosenkov Can you check whether setting the toolchain via env var still works for you in this branch? It should, but the knitr test I added may fail. Not sure.

It seems to work. If it is OK on the CI I guess we should not worry.

**Comment by clauswilke** [comment]
The env var feature is not being used on CI, though, is it?

File: R/eval.R (hunk 1 above)
Comment (clauswilke) [reply R/eval.R:36]: I'll make that change.

**Comment by Ilia-Kosenkov** [comment]
> The env var feature is not being used on CI, though, is it?

It is used in extendr now, to point rextendr to the local (and the most recent) version of extendr-api.

**Comment by clauswilke** [comment]
Are you certain that it actually does that, though? I tried the following and the check ran just fine:
```r
Sys.setenv(REXTENDR_TOOLCHAIN = "xyz")
rcmdcheck::rcmdcheck(".")
```
As far as I can tell the env var doesn't make it into the check in this way. It may have to be written into `.Renviron`.

In any case, I'm going to merge this PR and then we can go from there.

**Comment by Ilia-Kosenkov** [comment]
@clauswilke,
Very strange. It fails on Windows
<details>

```

R version 4.0.3 (2020-10-10) -- "Bunny-Wunnies Freak Out"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(testthat)
> library(rextendr)
> 
> # no need to run tests if cargo isn't installed.
> cargo_available <- system2("cargo") == 0L
Rust's package manager

USAGE:
    cargo [+toolchain] [OPTIONS] [SUBCOMMAND]

OPTIONS:
    -V, --version           Print version info and exit
        --list              List installed commands
        --explain <CODE>    Run `rustc --explain CODE`
    -v, --verbose           Use verbose output (-vv very verbose/build.rs output)
    -q, --quiet             No output printed to stdout
        --color <WHEN>      Coloring: auto, always, never
        --frozen            Require Cargo.lock and cache are up to date
        --locked            Require Cargo.lock is up to date
        --offline           Run without accessing the network
    -Z <FLAG>...            Unstable (nightly-only) flags to Cargo, see 'cargo -Z help' for details
    -h, --help              Prints help information

Some common cargo commands are (see all commands with --list):
    build, b    Compile the current package
    check, c    Analyze the current package and report errors, but don't build object files
    clean       Remove the target directory
    doc         Build this package's and its dependencies' documentation
    new         Create a new cargo package
    init        Create a new cargo package in an existing directory
    run, r      Run a binary or example of the local package
    test, t     Run the tests
    bench       Run the benchmarks
    update      Update dependencies listed in Cargo.lock
    search      Search registry for crates
    publish     Package and upload this package to the registry
    install     Install a Rust binary. Default location is $HOME/.cargo/bin
    uninstall   Uninstall a Rust binary

See 'cargo help <command>' for more information on a specific command.
> not_cran <- identical(Sys.getenv("NOT_CRAN"), "true")
> 
> if (cargo_available && not_cran) {
+   test_check("rextendr")
+ }
>> {rextendr}: Using toolchain from 'REXTENDR_TOOLCHAIN': xyz
    Updating git repository `https://github.com/extendr/extendr`
    Updating crates.io index
   Compiling winapi-build v0.1.1
   Compiling winapi-x86_64-pc-windows-gnu v0.4.0
   Compiling winapi v0.3.9
   Compiling winapi v0.2.8
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.60
   Compiling extendr-engine v0.1.11 (https://github.com/extendr/extendr#f2352167)
   Compiling lazy_static v1.4.0
   Compiling kernel32-sys v0.2.2
   Compiling quote v1.0.8
   Compiling extendr-macros v0.1.11 (https://github.com/extendr/extendr#f2352167)
   Compiling libR-sys v0.2.1
   Compiling extendr-api v0.1.11 (https://github.com/extendr/extendr#f2352167)
   Compiling rextendr1 v0.0.1 (C:\Users\[redacted]\AppData\Local\Temp\RtmpeO5AWU\file22c824226064)
    Finished dev [unoptimized + debuginfo] target(s) in 34.02s
    Updating crates.io index
   Compiling rextendr2 v0.0.1 (C:\Users\[redacted]\AppData\Local\Temp\RtmpeO5AWU\file22c824226064)
    Finished dev [unoptimized + debuginfo] target(s) in 1.40s
error: toolchain 'xyz' is not installed
error: toolchain 'xyz' is not installed
== Failed tests ================================================================
-- Error (test-name-override.R:19:3): Multiple rust functions with the same name --
Error: Rust code could not be compiled successfully. Aborting.
Backtrace:
    x
 1. \-rextendr::rust_source(code = rust_src_1, quiet = FALSE) test-name-override.R:19:2
-- Error (test-source.R:25:3): `rust_source()` works ---------------------------
Error: Rust code could not be compiled successfully. Aborting.
Backtrace:
    x
 1. \-rextendr::rust_source(code = rust_src, quiet = FALSE, cache_build = TRUE) test-source.R:25:2

[ FAIL 2 | WARN 0 | SKIP 0 | PASS 4 ]
Error: Test failures
Execution halted

```

</details>

But somehow ignores this value on Linux (WSL). I need to investigate.

**Comment by Ilia-Kosenkov** [comment]
I bet you do not have `NOT_CRAN` set to string value `"true"` (all lower case).
Tests are ignored 'on CRAN'.
I have hard-coded this env var in my Windows, but not in WSL, so now WSL does its job:
<details>

```

R version 4.0.3 (2020-10-10) -- "Bunny-Wunnies Freak Out"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(testthat)
> library(rextendr)
> 
> # no need to run tests if cargo isn't installed.
> cargo_available <- system2("cargo") == 0L
Rust's package manager

USAGE:
    cargo [+toolchain] [OPTIONS] [SUBCOMMAND]

OPTIONS:
    -V, --version           Print version info and exit
        --list              List installed commands
        --explain <CODE>    Run `rustc --explain CODE`
    -v, --verbose           Use verbose output (-vv very verbose/build.rs output)
    -q, --quiet             No output printed to stdout
        --color <WHEN>      Coloring: auto, always, never
        --frozen            Require Cargo.lock and cache are up to date
        --locked            Require Cargo.lock is up to date
        --offline           Run without accessing the network
    -Z <FLAG>...            Unstable (nightly-only) flags to Cargo, see 'cargo -Z help' for details
    -h, --help              Prints help information

Some common cargo commands are (see all commands with --list):
    build, b    Compile the current package
    check, c    Analyze the current package and report errors, but don't build object files
    clean       Remove the target directory
    doc         Build this package's and its dependencies' documentation
    new         Create a new cargo package
    init        Create a new cargo package in an existing directory
    run, r      Run a binary or example of the local package
    test, t     Run the tests
    bench       Run the benchmarks
    update      Update dependencies listed in Cargo.lock
    search      Search registry for crates
    publish     Package and upload this package to the registry
    install     Install a Rust binary. Default location is $HOME/.cargo/bin
    uninstall   Uninstall a Rust binary

See 'cargo help <command>' for more information on a specific command.
> not_cran <- identical(Sys.getenv("NOT_CRAN"), "true")
> 
> if (cargo_available && not_cran) {
+   test_check("rextendr")
+ }
>> {rextendr}: Using toolchain from 'REXTENDR_TOOLCHAIN': xyz
    Updating git repository `https://github.com/extendr/extendr`
    Updating crates.io index
 Downloading crates ...
  Downloaded syn v1.0.60
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling libR-sys v0.2.1
   Compiling syn v1.0.60
   Compiling extendr-engine v0.1.11 (https://github.com/extendr/extendr#f2352167)
   Compiling lazy_static v1.4.0
   Compiling quote v1.0.8
   Compiling extendr-macros v0.1.11 (https://github.com/extendr/extendr#f2352167)
   Compiling extendr-api v0.1.11 (https://github.com/extendr/extendr#f2352167)
   Compiling rextendr1 v0.0.1 (/tmp/RtmpJB52ul/file48a3137219e2)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 22s
    Updating crates.io index
   Compiling rextendr2 v0.0.1 (/tmp/RtmpJB52ul/file48a3137219e2)
    Finished dev [unoptimized + debuginfo] target(s) in 0.94s
error: toolchain 'xyz' is not installed
── ERROR (test-name-override.R:19:3): Multiple rust functions with the same name
Error: Rust code could not be compiled successfully. Aborting.
Backtrace:
    █
 1. └─rextendr::rust_source(code = rust_src_1, quiet = FALSE) test-name-override.R:19:2

error: toolchain 'xyz' is not installed
── ERROR (test-source.R:25:3): `rust_source()` works ───────────────────────────
Error: Rust code could not be compiled successfully. Aborting.
Backtrace:
    █
 1. └─rextendr::rust_source(code = rust_src, quiet = FALSE, cache_build = TRUE) test-source.R:25:2

══ testthat results  ═══════════════════════════════════════════════════════════
ERROR (test-name-override.R:19:3): Multiple rust functions with the same name
ERROR (test-source.R:25:3): `rust_source()` works

[ FAIL 2 | WARN 0 | SKIP 0 | PASS 4 ]
Error: Test failures
Execution halted

```

</details>

**Comment by Ilia-Kosenkov** [comment]
@clauswilke,
I am not sure how `knitr` does this, but it seems it ignores the override of formals, see passed tests.

**Comment by clauswilke** [comment]
It has nothing to do specifically with knitr, it's because it uses `rust_eval()` which then calls `rust_source()`. Since that call was read when the package was built, changing the formals doesn't reach it. I suspected this would be a problem. I've added a test for `rust_eval()` and you should see the same happening for that test.

We'll probably have to go to `options()` after all.

**Comment by clauswilke** [comment]
But now something unrelated has happened and the entire CI build is broken.

**Comment by Ilia-Kosenkov** [comment]
@clauswilke,
It looks like something transient, I have restarted CI here and in my fork -- seems to work again.
Speaking of markdown issue - perhaps we can leave it as is for now, until we implement options & other things.

**Pull request merged**